File: Prolog/part_3/mama_mia.pl
% /* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
% 
% * File Name : mama_mia.pl
% 
% * Purpose :
% 
% * Creation Date : 28-06-2011
% 
% * Last Modified : Tue 05 Jul 2011 03:48:48 AM EEST
% 
% * Created By : Greg Liras <gregliras@gmail.com>
% 
% _._._._._._._._._._._._._._._._._._._._._.*/

  execuTe(_,_,I,OM,""):- OM=I.
  execuTe(_,M,I,OM,"M"):- OM is M*I.
  execuTe(A,_,I,OM,"A"):- OM is A+I.

  big_sista(_,_,OMO,OMO,[]).
  big_sista(A,M,I  ,O  ,[P|Prog]):-
    execuTe(A,M,I,I2,P),
    big_sista(A,M,I2,O,Prog).

  big_mama(A,M,LI,HI,LO,HO,Prog,RetProg):-
    big_sista(A,M,LI,MLO,Prog),
    big_sista(A,M,HI,MHO,Prog),
    (
      MLO >= LO,
      MLO =< HO,
      MHO >= LO,
      MHO =< HO -> RetProg = Prog
      ;
        (MLO > HO ; MHO < LO) -> fail
      ;
        !,
        moreProg(Prog,RProg),
        big_mama(A,M,LI,HI,LO,HO,RProg,RetProg)
    ).

  moreProg(IProg,OProg):-OProg=["A"|IProg];OProg=["M"|IProg].

  mama_mia(A,M,LI,HI,LO,HO,Prog):-
    big_mama(A,M,LI,HI,LO,HO,[],Prog).

File: Java/part_3/MamaMia.java
/* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.

* File Name : MamaMia.java

* Purpose :

* Creation Date : 05-07-2011

* Last Modified : Wed 03 Aug 2011 02:29:03 PM EEST

* Created By : Greg Liras <gregliras@gmail.com>

_._._._._._._._._._._._._._._._._._._._._.*/

import java.util.ArrayList;


public class MamaMia
{
  public static void main(String[] input)
  {
    if(input.length != 6)
    {
      System.out.println("Usage: java MamaMia a m li hi lo ho");
    }
    else
    {
      int a = Integer.parseInt(input[0]);
      int m = Integer.parseInt(input[1]);
      int li = Integer.parseInt(input[2]);
      int hi = Integer.parseInt(input[3]);
      int lo = Integer.parseInt(input[4]);
      int ho = Integer.parseInt(input[5]);
      int minLimit = Math.min((int) Math.floor(lo/a)
                 ,(int) (Math.floor(Math.log(lo)/Math.log(m))))-1;
      int maxLimit = Math.max((int) Math.ceil(ho/a)
                 ,(int) (Math.ceil(Math.log(ho)/Math.log(m))));

      int status=-1;
      ProgramsGenerator PGen = new ProgramsGenerator(minLimit);
      ArrayList<MBitSet> Progs=null;
      ArrayList<MBitSet> newProgs=null;
      Runner runner=null;

      runner = new Runner();
      newProgs = new ArrayList<MBitSet>();
      PGen.fillProgList();
      for(int i = 0 ; i<= maxLimit ; i++)
      {
        Progs = PGen.getProgList();
        newProgs.clear();
        newProgs.ensureCapacity(Progs.size());
        for(MBitSet Prog:Progs)
        {
          runner.SetRunner(a,m,li,hi,Prog);
          runner.run();
          status = runner.outPutCheck(lo,ho);
          if(status==0)
          {
            if(Progs.size()==1)
            {
              System.out.println("empty");
            }
            else
            { 
              System.out.println(Prog);
            }
            return;
          }
          else if (status==1)
          {
            //System.out.println("Status 1");
          }
          else if (status==2)
          {
            newProgs.add(Prog);
            //System.out.println("Status 2");
          }
        //System.out.println(Prog.toString().length()+" "+Progs.size());
        }
        //PGen.printProgsSize();
        newProgs.trimToSize();
        PGen.setProgList(newProgs);
        PGen.makeMoreProgs();
      }
      System.out.println("impossible");
    }
  }
}
